dashbuilder-backend/dashbuilder-navigation-backend/src/main/java/org/dashbuilder/navigation/service/LayoutTemplateAnalyzer.java

```diff
@@ -43,44 +50,115 @@ public LayoutTemplateAnalyzer(PerspectivePluginServicesImpl pluginServices, Navi
-            boolean deadlock = hasDeadlock(row, ancestorSet);
+            boolean deadlock = analyzeDeadlock(row, deadlockInfo, navGroupId);
             if (deadlock) {
                 return true;
             }
         }
         return false;
     }
 
-    public boolean hasDeadlock(LayoutRow row, Set<String> ancestorSet) {
+    public boolean analyzeDeadlock(LayoutRow row, LayoutDeadlockInfo deadlockInfo, String rootGroupId) {
         NavTree navTree = navigationServices.loadNavTree();
         for (LayoutColumn column : row.getLayoutColumns()) {
 
             for (LayoutComponent component : column.getLayoutComponents()) {
+                deadlockInfo.push(new LayoutNavigationRef(NAV_COMPONENT, component.getDragTypeName()));
 
-                // Any layout component linked to a perspective can potentially lead to a deadlock scenario.
-                String perspectiveId = component.getProperties().get("perspectiveId");
+                // Components pointing to a perspective can cause a deadlock
+                String perspectiveId = component.getProperties().get(PERSPECTIVE_ID);
                 if (perspectiveId != null) {
-                    boolean deadlock = hasDeadlock(perspectiveId, ancestorSet);
+                    boolean deadlock = analyzeDeadlock(perspectiveId, deadlockInfo);
+                    if (deadlock) {
+                        return true;
+                    }
+                }
+
+                // Get the nav group the component is tied to
+                String navGroupId = component.getProperties().get(NAV_GROUP_ID);
+                String navDefaultId = component.getProperties().get(NAV_DEFAULT_ID);
+                LayoutNavigationRefType navGroupRefType = NAV_GROUP_DEFINED;
+                if (rootGroupId != null) {
+                    navGroupId = rootGroupId;
+                    navDefaultId = null;
+                    navGroupRefType = NAV_GROUP_CONTEXT;
+                }
+                else {
+                    LayoutNavigationRef lastDefaultItemRef = deadlockInfo.getLastDefaultItemRef();
+                    if (lastDefaultItemRef != null) {
+                        NavItem lastDefaultItem = navTree.getItemById(lastDefaultItemRef.getName());
+                        String lastDefaultGroupId = NavWorkbenchCtx.get(lastDefaultItem).getNavGroupId();
+                        if (lastDefaultGroupId != null) {
+                            navGroupId = lastDefaultGroupId;
+                            navDefaultId = null;
+                            navGroupRefType = NAV_GROUP_CONTEXT;
+                        }
+                    }
+                }
+                NavGroup navGroup = (NavGroup) navTree.getItemById(navGroupId);
+
+                // The configured default item can cause a deadlock
+                if (navDefaultId != null) {
+                    deadlockInfo.push(new LayoutNavigationRef(navGroupRefType, navGroup.getId()));
```

Comment:
why push twice?

Reply:
It is the context of the recursion analysis. We need to push both elements, group and default item so that we can report what elements are creating the recursion issue.